### PR TITLE
default to v2 of sst version on the write path

### LINKED ISSUE #1210: Raise key size limit from u16 to u32
**Is your feature request related to a problem? Please describe.**
In some use cases (e.g. GROUP BY in Flink), key is an arbitrary set of records specified by the user.
Therefore, it might exceed the current limit of 65Kb, leading to a panic.

**Describe the solution you'd like**

1. Add a new `SstRowCodecV1` implementation using `u32` instead of `u16` for key length
2. Use this new version by default
3. Change the related assertions from `u16` to `u32`

Migration shouldn't be necessary as long as we read SST in whatever format it was written in; and write in a new format

**Describe alternatives you've considered**
Delta encoding was proposed by @agavra; while being presumably more performant, it's also more complex.

**Additional context**
Discussion thread: https://discord.com/channels/1232385660460204122/1232385660946616433/1463192563799556288

### PR DESCRIPTION
**Pull request by agavra**
## Summary

Fixes #1210 

This PR wires in #1237 and #1248 into the default read path for SlateDB. V1 SSTs can no longer be created except for under `#[cfg(test)]` (though we can debate if we want to expose that publicly) though they can still be read (see tests to prove it).

This test also fixes a bug with many versions of the same key straddling restart points.

## Changes

(Change summary generated by Mr. Claude and Mrs. Claudette)

- **Default SST format changed to V2** - `SstBuilder` now defaults to `BlockFormat::V2` with prefix compression instead of V1

- **Bug fix in `BlockIteratorV2::find_restart_for_key`** - When binary search finds a restart point with key exactly equal to target, it now returns the *previous* restart point to avoid missing duplicate key entries that might exist before the found restart

- **`last_written_key_and_seq` now version-aware** - Branches on SST version to use `BlockIterator` (V1) or `BlockIteratorV2` (V2) for reading the last entry

- **Added `with_block_format()` builder method** - Test-only API on `DbBuilder` and `SstBuilder` to force V1 format for backward compatibility testing

- **Test coverage for duplicate keys across restart points** - 3 new tests covering edge cases where duplicate keys span multiple restart regions

- **Test for V1 backward compatibility** - New test `should_get_last_written_key_and_seq_from_v1_sst` ensures V1 path still works

## Notes for Reviewers

There are a few minor changes coupled into this to make it work, but the PR remains pretty small. I'm happy to break it up if requested.

## Checklist

- [X] Small, scoped PR (< 500 total lines excluding tests); or opened as Draft with a plan on how to break it into smaller pieces
- [X] Linked related issue(s) or added context in the description
- [X] Self-reviewed the diff; added comments for tricky parts
- [X] Tests added/updated and passing locally
- [X] Ran `cargo fmt`, `cargo clippy --all-targets --all-features`, and `cargo nextest run --all-features`
- [X] Called out any breaking changes and provided migration notes
- [X] Considered performance impact; added notes or benchmarks if relevant

Thank you for the review! 🙏

### CODE REVIEW

File: slatedb/src/format/sst.rs [hunk 1]
```diff
@@ -497,23 +500,24 @@ impl Default for SsTableFormat {
...
             compression_codec: None,
             block_transformer: None,
+            block_format: None,
         }
     }
 }
 
 impl SsTableFormat {
-    async fn read_footer_header(
+    async fn read_metadata_offset_and_version(
```
Comment (agavra) [inline slatedb/src/format/sst.rs:514]: addresses a comment from a previous PR that pointed out `(u64, u16)` was a confusing return type. this should hopefully clear that up

File: slatedb/src/block_iterator_v2.rs [hunk 2]
```diff
@@ -130,27 +127,42 @@ impl<B: BlockLike> AscendingState<B> {
         self.exhausted || self.offset_in_block >= self.block.data().len()
     }
 
-    /// Returns the largest restart index where the key at that restart point is <= target.
+    /// Returns the restart index to start scanning from when seeking to target.
+    ///
+    /// When a restart point has key exactly equal to target, we return the PREVIOUS
+    /// restart point to ensure we don't miss any entries with that key that might
+    /// exist before the found restart point.
     fn find_restart_for_key(&self, target: &[u8]) -> usize {
```
Comment (agavra) [inline slatedb/src/block_iterator_v2.rs]: this is the bug fix

File: slatedb/src/sst_iter.rs [hunk 3]
```diff
@@ -1247,7 +1247,7 @@ mod tests {
             .unwrap();
         let sst_handle = table_store.open_sst(&SsTableId::Wal(0)).await.unwrap();
         let index = table_store.read_index(&sst_handle, true).await.unwrap();
-        assert_eq!(index.borrow().block_meta().len(), 10);
+        assert_eq!(index.borrow().block_meta().len(), 8);
```
Comment (agavra) [inline slatedb/src/sst_iter.rs:1250]: V2 blocks are a bit smaller because they use fewer offsets

File: slatedb/src/utils.rs [hunk 4]
```diff
@@ -194,12 +196,27 @@ pub(crate) async fn last_written_key_and_seq(
 
     // Sort descending so we get the last row from the last block, which
     // should be the last written key/seq.
-    let mut block_iter = BlockIterator::new(block, IterationOrder::Descending);
-    block_iter.init().await?;
-    Ok(block_iter
-        .next_entry()
-        .await?
-        .map(|entry| (entry.key, entry.seq)))
+    let sst_version = table_store.read_sst_version(output_sst).await?;
```
Comment (agavra) [inline slatedb/src/utils.rs:199]: this is another "fix" which I had missed from before

File: slatedb/src/wal/wal_sst_builder.rs [hunk 5]
```diff
@@ -276,18 +276,17 @@ impl EncodedWalSsTableBuilder {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::block_iterator::BlockIterator;
+    use crate::block_iterator_v2::BlockIteratorV2;
```
Comment (criccomini) [inline slatedb/src/wal/wal_sst_builder.rs]: Do you think it makes sense to have a `BlockIteratorLatest` or something so future changes don't require a bunch of V2->V3? updates as well? There seem to be quite a few.

Same question for `SST_FORMAT_VERSION_LATEST` and `BlockBuilder::new_latest`?

File: slatedb/src/db/builder.rs [hunk 6]
```diff
@@ -198,9 +199,18 @@ impl<P: Into<Path>> DbBuilder<P> {
...
             #[cfg(feature = "compaction_filters")]
             compaction_filter_supplier: None,
+            block_format: None,
         }
     }
 
+    /// Sets the block format for SST files. This is only available in tests
+    /// to verify backward compatibility between V1 and V2 formats.
+    #[cfg(test)]
+    pub fn with_block_format(mut self, block_format: crate::sst_builder::BlockFormat) -> Self {
```
Comment (criccomini) [inline slatedb/src/db/builder.rs]: This should be a `Setting` in `config.rs` since it's a config not a component (i.e. dependency injection). Let's leave it a `cfg(test)` for now, still.

File: slatedb/src/sst_builder.rs [hunk 7]
```diff
@@ -1226,4 +1231,126 @@ mod tests {
...
+        let mut iter = SstIterator::new_owned_initialized(
+            ..,
+            sst_handle,
+            Arc::new(table_store),
+            SstIteratorOptions::default(),
+        )
+        .await
+        .unwrap()
+        .unwrap();
+
```
Comment (criccomini) [inline slatedb/src/sst_builder.rs:1323]: Nit: might be worth verifying that the stored data is actually using the V1 format.

File: slatedb/src/sst_builder.rs [hunk 8]
```diff
@@ -1226,4 +1231,126 @@ mod tests {
...
```
Comment (criccomini) [inline slatedb/src/sst_builder.rs:1386]: Nit: same here

File: slatedb/src/block_iterator_v2.rs [hunk 9]
```diff
@@ -130,24 +127,79 @@ impl<B: BlockLike> AscendingState<B> {
         self.exhausted || self.offset_in_block >= self.block.data().len()
     }
 
-    /// Returns the largest restart index where the key at that restart point is <= target.
-    fn find_restart_for_key(&self, target: &[u8]) -> usize {
+    /// Returns the restart index to start scanning from when seeking to target.
+    ///
+    /// When a restart point has key exactly equal to target, we return the PREVIOUS
```
Comment (criccomini) [inline slatedb/src/block_iterator_v2.rs:132]: Is it safe to return only the previous restart point? What if there are so many rows for the same key that it ends up with multiple restart points all pointing to the same key? IIRC, we have 16 rows/restart point, so if the same key takes up > 32 rows, we would need to go back even farther (IIUC).

File: slatedb/src/block_iterator_v2.rs [hunk 10]
```diff
@@ -1092,4 +1144,232 @@ mod tests {
         let kv = iter.next_entry().await.unwrap().unwrap();
         assert_eq!(kv.key.as_ref(), b"key_09");
     }
+
+    // ==================== Duplicate Key Edge Case Tests ====================
+
```
Comment (criccomini) [inline slatedb/src/block_iterator_v2.rs:1149]: A proptest in here that really exercises a wide array of possibilities would make me a little more comfortable if it's not a ton of trouble.

File: slatedb/src/wal/wal_sst_builder.rs (hunk 5 above)
Comment (agavra) [reply slatedb/src/wal/wal_sst_builder.rs]: Love it. Will make that change.

File: slatedb/src/block_iterator_v2.rs (hunk 9 above)
Comment (agavra) [reply slatedb/src/block_iterator_v2.rs:132]: the code accounts for that, it finds the first/last (for ascending/descending) restart point that has the key we're looking for. if that is exactly equal to the key, we need to go yet another one before/after in case the last/first key of the of the previous range is the value we're looking for.

that was a bit confusing, but check out the tests it should cover this.

File: slatedb/src/block_iterator_v2.rs (hunk 10 above)
Comment (agavra) [reply slatedb/src/block_iterator_v2.rs:1149]: > if it's not a ton of trouble

_asks Claude to roll up sleeves_ 😆 in today's day and age, asking for more tests is never a ton of trouble and always welcome.

File: slatedb/src/block_iterator_v2.rs (hunk 9 above)
Comment (agavra) [reply slatedb/src/block_iterator_v2.rs:132]: note this part of the rustdoc: "Binary search for the **first restart index** where key >= target." (emphasis added)

Comment (criccomini) [reply slatedb/src/block_iterator_v2.rs:132]: Gotcha. Yea, I saw the test checked multiple offsets, but only two. I was kinda curious how it behaved with 3+ (hence the proptest request below). I couldn't quite figure out the logic in the search, though I noticed you removed the `==`, so I guess that's what shifts the binary search?

File: slatedb/src/wal/wal_sst_builder.rs [hunk 11]
```diff
@@ -187,7 +188,7 @@ impl EncodedWalSsTableBuilder {
             return Ok(None);
         }
 
-        let new_builder = BlockBuilder::new_v1(self.block_size_config);
+        let new_builder = BlockBuilder::new_v2(self.block_size_config);
```
Comment (criccomini) [inline slatedb/src/wal/wal_sst_builder.rs]: Is this intended?

```suggestion
        let new_builder = BlockBuilder::new_latest(self.block_size_config);
```

File: slatedb/src/sst_builder.rs [hunk 12]
```diff
@@ -852,7 +859,7 @@ mod tests {
 
         // then:
         for expected_entries in expected_blocks {
-            let mut iter = BlockIterator::new(blocks.pop_front().unwrap(), Ascending);
+            let mut iter = BlockIteratorV2::new_ascending(blocks.pop_front().unwrap());
```
Comment (criccomini) [inline slatedb/src/sst_builder.rs]: IDK how to best handle this, but I'm wondering if `BlockIteratorLatest` for these tests would be better? It seems every time we bump the iterator version, we'll no longer be testing the older iterators. But I suppose that's OK provided we don't modify them later. 🤔

File: slatedb/src/wal/wal_sst_builder.rs (hunk 11 above)
Comment (agavra) [reply slatedb/src/wal/wal_sst_builder.rs]: I was debating whether production code should also automatically switch over, I guess there's no reason why not. I updated this and the `sst_builder` (which also hardcoded V2 as the format) in the next commit.

File: slatedb/src/sst_builder.rs (hunk 12 above)
Comment (agavra) [reply slatedb/src/sst_builder.rs]: Yeah, I'll change this one as well. I think as long as we still have tests for old block formats and some mixed version tests to make sure SSTs can read old/new then that's likely sufficient coverage.

**Comment by agavra** [comment]
Merged. @criccomini I'll try to keep an eye on the benchmarks but i suspect we might say a bit of a regression in scenarios that can cache most of the workload. we'll see improvements in situations where the workload is not fully cached. IIRC, the benchmark hits cache nearly 100% of the time.
